### PR TITLE
chore: homogenising promotion

### PR DESCRIPTION
**Pull request by FauxFaux**
Follows #36.

A binary version of `impl_array`?

`promotion` now returns a `Pair` of identical arrays, and then maths operators can be implemented on `Pair`, instead of everywhere?

`v_starco` and `v_idot` also want to do this, but need to handle `chars`, and maybe don't want promotions? Maybe they do and you couldn't face it? Maybe we need a non-promoting version of homogenisation, which returns a non-maths pair, or we need to go back to a type of `char` which implements some maths operators?

Or we could stop using ops::Add::add and just "define" addition on char using.. some other trait monstrosity.

This PR doesn't actually reduce lines of code much, but it feels "correct" to me, the macros have moved "into" the enums (even though it's a new enum), and away from the "application code" (in `v_plus`)?

### CODE REVIEW

**Comment by AshyIsMe** [comment]
Brilliant, was editing an in depth comment and clicked the wrong link and the browser ate it...

TLDR: gonna merge this
